gcloud-java-examples/src/main/java/com/google/gcloud/examples/DnsExample.java

```diff
@@ -0,0 +1,516 @@
+          new ListDnsRecordsAction().run(dns, args);
+        }
+        if (args.length == 1 || "changes".equals(args[1])) {
+          new ListChangesAction().run(dns, args);
+        }
+      }
+    }
+
+    @Override
+    public boolean check(String... args) {
+      if (args.length == 0 || args.length == 1) {
+        return true;
+      }
+      if ("records".equals(args[1])) {
+        return new ListDnsRecordsAction().check(args);
+      }
+      if ("changes".equals(args[1])) {
+        return new ListChangesAction().check(args);
+      }
+      return false;
+    }
+
+    @Override
+    public String params() {
+      return "[<zone_name> [changes [descending | ascending] | records]]";
+    }
+  }
+
+  private static class GetProjectAction implements DnsAction {
+
+    @Override
+    public void run(Dns dns, String... args) {
+      ProjectInfo project = dns.getProject();
+      System.out.printf("Project id: %s%nQuota:%n", dns.options().projectId());
+      System.out.printf("\tZones: %d%n", project.quota().zones());
+      System.out.printf("\tDNS records per zone: %d%n", project.quota().rrsetsPerZone());
+      System.out.printf("\tRecord sets per DNS record: %d%n",
+          project.quota().resourceRecordsPerRrset());
+      System.out.printf("\tAdditions per change: %d%n", project.quota().rrsetAdditionsPerChange());
+      System.out.printf("\tDeletions per change: %d%n", project.quota().rrsetDeletionsPerChange());
+      System.out.printf("\tTotal data size per change: %d%n",
+          project.quota().totalRrdataSizePerChange());
+    }
+
+    @Override
+    public String params() {
+      return "";
+    }
+
+    @Override
+    public boolean check(String... args) {
+      return args.length == 0;
+    }
+  }
+
+  static {
+    ACTIONS.put("create", new CreateZoneAction());
+    ACTIONS.put("delete", new DeleteZoneAction());
+    ACTIONS.put("get", new GetZoneAction());
+    ACTIONS.put("list", new ListAction());
+    ACTIONS.put("add-record", new AddDnsRecordAction());
+    ACTIONS.put("delete-record", new DeleteDnsRecordAction());
+    ACTIONS.put("quota", new GetProjectAction());
+  }
+
+  private static void printUsage() {
+    StringBuilder actionAndParams = new StringBuilder();
+    for (Map.Entry<String, DnsAction> entry : ACTIONS.entrySet()) {
+      actionAndParams.append('\t').append(System.lineSeparator()).append(entry.getKey());
+      String param = entry.getValue().params();
+      if (param != null && !param.isEmpty()) {
+        actionAndParams.append(' ').append(param);
+      }
+    }
+    System.out.printf("Usage: %s [<project_id>] operation <args>*%s%n",
+        DnsExample.class.getSimpleName(), actionAndParams);
+  }
+
+  public static void main(String... args) throws Exception {
+    if (args.length < 1) {
+      System.out.println("Missing required action");
+      printUsage();
+      return;
+    }
+    DnsOptions.Builder optionsBuilder = DnsOptions.builder();
```

Comment:
Nit: instead of keeping the `DnsOptions` around when we could exit due to invalid input, you could just store the project ID here and use it to create the service in the same "if" block as in line 510